### PR TITLE
react-relay and relay-runtime: Update TS definitions for v11 RC

### PR DESCRIPTION
**Pull request by rbalicki2**
An RC for v11 (which stabilizes hooks) will come out tomorrow. This PR adds types to the v10 to reflect the v11 RC. Unfortunately, DefinitelyTyped doesn't support versions of the form v11.0.0-rc.1, so we chose to update the v10 types to reflect this change.

The breaking changes in the RC (in brief) are:

* relay-runtime: fetchQuery (which returns a promise) is renamed fetchQuery_DEPRECATED
* relay-runtime: fetchQuery (which returns an Observable) is exported in its place

Non-breaking changes in the RC are:
* Added: /hooks now exports the relay hooks APIs
* Added: /legacy now exports the Containers APIs
* Added: /index now also exports the relay hooks APIs.
I will update this post tomorrow to link to the full RC announcement.

This PR:
* Updates types to match the RC structure
* Re-uses existing types for the Containers APIs
* Re-uses existing types (i.e. from relay-experimental) for the hooks APIs
* Handles fetchQuery appropriately

-------

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: github.com/facebook/relay/issues (see the post about the RC. The post will come out 2/26)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by rbalicki2** [comment]
@alloy @maraisr @edvinerikson  ugh, sorry, I ended up reusing pretty much all the types. Will add y'all back as owners.

File: types/react-relay/tsconfig.json
Comment (maraisr) [inline types/react-relay/tsconfig.json:22]: think we may need to expose the /legacy and /hooks here (if those marry up to the export sub-modules of the packages)

**Review by maraisr — APPROVED** [review]
This was so good, thanks mate! I'm going to approve this on the bounds that its just a file-rejig. My thinking here is that now that we're entering this stuff into _stable something_ we should have more thorough tests, stuff that _could_ catch that `fetchQuery` to deprecated stuff how cacheConfig argument changes. But on the outset, think this is solid.

File: types/react-relay/tsconfig.json
Comment (rbalicki2) [reply types/react-relay/tsconfig.json:22]: Ah, yeah! Thanks for catching that.

I wonder if we should expose all of the files individually as well, since it is possible to import individual files.

Comment (rbalicki2) [reply types/react-relay/tsconfig.json:22]: Although, I wonder why that wasn't required for the previous set of types? The readme makes it seem like, since these files are accessible from index.d.ts, I should be good... fingers crossed.

Comment (rbalicki2) [reply types/react-relay/tsconfig.json:22]: I wonder if something changed. I get 

```
Only index.d.ts may be listed explicitly in tsconfig's "files" entry.
Other d.ts files must either be referenced through index.d.ts, tests, or added to OTHER_FILES.txt.
```

**Comment by rbalicki2** [comment]
Working on fixing these CI errors on another branch

**Comment by rbalicki2** [comment]
cc @maraisr CI builds are passing. Could you please re-review?

Changes since the earlier versions:
* test only on the latest; I couldn't get tests to work within v7
* add an empty UNUSED_FILES.txt file to v7 (I don't know why this is necessary)
* limit the included files in tsconfig.json. My guess, from looking at types/ol, is that these don't need to be explicitly exported

File: types/react-relay/legacy.d.ts
Comment (maraisr) [inline types/react-relay/legacy.d.ts:81]: @rbalicki2 can you just help me and check if this is correct? Struggling to understand the flow equivalent to this one

File: types/react-relay/test/relay-hooks-tests.tsx
Comment (maraisr) [inline types/react-relay/test/relay-hooks-tests.tsx:595]: @rbalicki2 correct that these are not exported? Couldn't find them in the hooks or anything.

**Review by maraisr — APPROVED** [review]
@rbalicki2 going to approve this one—there are a few comments, but think on the outset this should be good. Its in RC still, so lets get it out, and get some broader testers in. Also although I think this v7 thing will work, and the /hooks—need to push "_something_" out to test that ha.

File: types/react-relay/test/relay-hooks-tests.tsx
Comment (rbalicki2) [reply types/react-relay/test/relay-hooks-tests.tsx:595]: Correct. These aren't exported. They will be, when React with concurrent mode is available.

Might make sense to delete the types, and get them back from the history, or just leave them commented. /shrug

File: types/react-relay/legacy.d.ts
Comment (rbalicki2) [reply types/react-relay/legacy.d.ts:81]: # Flow

As far as I can tell, Props2 = $RelayProps<Props1, RelayProp> is generating a new type which is like Props1, except:
* Excluding `{relay: RelayProp}`, and
* Mapping all remaining props which are generated GraphQLTaggedLiterals to the value of their $refType field (preserving things like nullability and whether it's an array), and
* leaving all remaining props alone

I think!

So, it is used in e.g. `createContainerWithProps` to create a new component type (Component2) to which you pass Props2, i.e.:

* don't pass `relay`
* pass `$key` types (e.g. FooFragment_user$key, I think)

And which reverses this, i.e. adds `relay` and looks up fragments by their ref (via RelayModernFragmentSpecResolver), and calls `Component1` with `Props1`.

# Typescript

That looks like what is happening in Typescript as well, but it's hard to be entirely sure.

Maybe it would be good to check whether all props are correctly mapped? e.g. making sure that all `FragmentOrRegularProp` matches the behavior of $RelayProps.

In particular, I would make sure that if the array is nullable, FragmentOrRegularProp applies. That seems to not be covered by FragmentOrRegularProp. But it's not clear!

See this: https://www.typescriptlang.org/play?#code/FAFwngDgpgBAgjAvDA3gQwFwxAJwK5QC+A3MKJLAEJLwwA+MAbgPYCWAJqedDAMIA8AFQB8NQTCgAPEFAB27AM60A-KlYLMMWXgC2AIyg5CMLCnV6s2-YZJlwPACI0BlYVwD07mE4UALZngANuwwAMaBUGg4gWASrCC+hjAGahqWugZGMMw4qRZaGTYANMl4IDDxAORKeswJysp2FDAAojQAFAgMLBwAlBLScooq2PiwWABmaIEKUKRAA
Type E seems to indicate that A | void doesn't extend A, but Typescript gets confused with type D.

Nullable arrays don't show up in Relay, though, so I think you're probably good.

# Conclusion

Did this work previously, with no complaints? If this is broken, it seems like it would have tripped someone else up.

I would land this, and maybe add some tests as a follow up?

**Comment by maraisr** [comment]
@elibarzilay not sure who I cannot merge this, am a codeowner in the file, but can't merge here without @alloy's approval?

**Review by maraisr — APPROVED** [review]

**Comment by maraisr** [comment]
Not sure why these tests are failing so badly. @elibarzilay mind rerunning them I cannot.

**Comment by alloy** [comment]
> This PR adds types to the v10 to reflect the v11 RC. Unfortunately, DefinitelyTyped doesn't support versions of the form v11.0.0-rc.1, so we chose to update the v10 types to reflect this change.

So are these changes compatible with v10? If not, we should not merge this until a stable release is out, constraining DT type versions can already be enough of a hassle for people as it is.

> Not sure why these tests are failing so badly. @elibarzilay mind rerunning them I cannot.

The failures seem contained to another package. Are these green on master?

**Comment by chriserickson** [comment]
>So are these changes compatible with v10? If not, we should not merge this until a stable release is out, constraining DT type versions can already be enough of a hassle for people as it is.

v11.0.0 has been released, so that shouldn't be a concern.

**Comment by rbalicki2** [comment]
Just a heads up for when you pull, I rebased onto master (which seems to have fixed the CI issues), as well as changing v10 -> v11, now that v11 is released.

**Review by maraisr — APPROVED** [review]

**Comment by maraisr** [comment]
Ready to merge

File: types/react-relay/legacy.d.ts
Comment (robrichard) [inline types/react-relay/legacy.d.ts:30]: @rbalicki2 it looks like in previous versions of `@types/react-relay`, index.d.ts re-exported the concrete `Environment` from relay-runtime. See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4c251988d62d2f0d7bdbb5b2fcf2d7570f2d1da7/types/react-relay/index.d.ts#L80

Was it intended for this to change in `react-relay/legacy`? Is it expected to be a type error when a function's argument is typed as `Environment` imported from `relay-runtime` and is passed a relay environment typed as  `Environment` imported from `react-relay/legacy` (or `react-relay`)?

```
Argument of type 'Environment' is not assignable to parameter of type 'RelayModernEnvironment'.
  Type 'Environment' is missing the following properties from type 'RelayModernEnvironment': configName, revertUpdate, replaceUpdate
```

**Pull request merged**
